client-tools/Utils/ClientTaskQueue.h

```diff
@@ -241,22 +253,37 @@ inline bool ClientTaskQueue<JobData>::spawnWorkers(
 
 template<typename JobData>
 inline bool ClientTaskQueue<JobData>::isQueueEmpty() const noexcept {
-  bool isEmpty = false;
-  try {
-    MUTEX_LOCKER(lock, _jobsLock);
-    isEmpty = _jobs.empty();
-  } catch (...) {
+  CONDITION_LOCKER(lock, _jobsCondition);
+  return _jobs.empty();
+}
+
+template<typename JobData>
+inline bool ClientTaskQueue<JobData>::isQueueEmptyAndAllWorkersIdle()
+    const noexcept {
+  // acquire both locks here, so we can avoid data races
+  CONDITION_LOCKER(lock1, _jobsCondition);
+  CONDITION_LOCKER(lock2, _workersCondition);
```

Comment:
Suggest add try_lock for ConditionVariable(pthread and TryEnterCriticalSection (also I think better to use srwmutex)) and use std::lock to avoid possible deadlock